File: crates/asio-poc/src/timer.rs
use tokio::sync::mpsc::{error::TrySendError, Sender};

#[cxx::bridge]
mod ffi {
    // message type that is sent  between
    // async tasks
    pub struct Message {
        pub count: usize,
        pub msg: String,
    }

    unsafe extern "C++" {
        include!("cpp-asio-poc/include/timer.h");

        type TimerTask;

        fn newTimerTask(call: Box<RustSender>, interval: u64) -> *mut TimerTask;
        unsafe fn deleteTimerTask(task: *mut TimerTask);
        unsafe fn start(timer: *mut TimerTask);
    }

    extern "Rust" {
        type RustSender;

        pub fn try_send(sender: &RustSender, msg: Message);
    }
}

pub struct RustSender(Sender<ffi::Message>);

impl RustSender {
    pub fn new(sender: Sender<ffi::Message>) -> Self {
        Self(sender)
    }
}

pub struct TimerTask(*mut ffi::TimerTask);

unsafe impl Send for TimerTask {}

impl TimerTask {
    pub fn new(sender: Sender<ffi::Message>, interval: u64) -> Self {
        let sender = Box::new(RustSender::new(sender));
        let ptr = ffi::newTimerTask(sender, interval);
        Self(ptr)
    }

    // might need self to be pinned?
    pub fn start(&self) {
        unsafe { ffi::start(self.0) }
    }
}

impl Drop for TimerTask {
    fn drop(&mut self) {
        unsafe {
            ffi::deleteTimerTask(self.0);
        }
    }
}

pub fn try_send(sender: &RustSender, msg: ffi::Message) {
    match sender.0.try_send(msg) {
        Ok(_) => {}
        Err(TrySendError::Full(_)) => {
            log::warn!("Channel full ignoring");
        }
        Err(TrySendError::Closed(_)) => log::error!("Channel closed!"),
    }
}

pub use ffi::Message;

File: crates/asio-poc/src/timer.cpp
#include "cpp-asio-poc/include/timer.h"
#include "cpp-asio-poc/include/io_context.h"

#include "cpp-asio-poc/src/timer.rs"

#include <cstdint>
#include <iostream>

class TimerTask::Impl
{
public:
  Impl(long unsigned interval, rust::Box<RustSender> call);
  void start();
  void cancel();

private:
  void print();

  asio::steady_timer timer_;
  rust::Box<RustSender> call_;
  long unsigned count;
};

TimerTask::Impl::Impl(long unsigned interval, rust::Box<RustSender> call)
    : call_(std::move(call)), count(0),
      timer_(get_io_instance(), asio::chrono::seconds(interval)) {}

void TimerTask::Impl::start()
{
  timer_.async_wait(std::bind(&TimerTask::Impl::print, this));
}
void TimerTask::Impl::cancel()
{
  // This method is tagged as deprecated so user should use non_error_code overload?? what that means?
  timer_.cancel();
}

void TimerTask::Impl::print()
{

  count += 1;

  // timer_.async_wait(std::bind(&TimerTask::Impl::print, this));
  timer_.async_wait([&, this](const asio::error_code &error)
                    {
    if (error == asio::error::operation_aborted)  
        return;
    std::string msg("From Cpp async task!!");
    Message message = {.count = count, .msg = msg};
    try_send(*call_, message);
    timer_.expires_at(timer_.expiry() + asio::chrono::seconds(1));
    this->print(); });
}

TimerTask::TimerTask(long unsigned int interval, rust::Box<RustSender> call)
    : impl(new class TimerTask::Impl::Impl(interval, std::move(call))) {}

void TimerTask::start()
{
  // safe as inner object is different than null
  this->impl->start();
}

void TimerTask::cancel()
{
  this->impl->cancel();
}

TimerTask *newTimerTask(rust::Box<RustSender> call, long unsigned interval)
{
  auto timerTask = new TimerTask(interval, std::move(call));
  return timerTask;
}

void deleteTimerTask(TimerTask *task)
{
  if (task != nullptr)
  {
    task->cancel();
    delete task;
  }
}

void start(TimerTask *timer)
{
  if (timer != nullptr)
  {
    timer->start();
  }
}
